packages/dev/core/test/unit/DeviceInput/babylon.deviceInput.test.ts

```diff
@@ -0,0 +1,226 @@
+    });
+
+    it("can use getDeviceSource", () => {
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const touchSource = new DeviceSource(deviceInputSystem, DeviceType.Touch, 0);
+        const touchSource2 = new DeviceSource(deviceInputSystem, DeviceType.Touch, 1);
+        const nullSource = deviceSourceManager.getDeviceSource(DeviceType.Touch, 0);
+        expect(nullSource).toBe(null);
+
+        deviceSourceManager._addDevice(touchSource);
+        deviceSourceManager._addDevice(touchSource2);
+
+        const firstAvailableSource = deviceSourceManager.getDeviceSource(DeviceType.Touch);
+        const specificSource = deviceSourceManager.getDeviceSource(DeviceType.Touch, 1);
+
+        expect(firstAvailableSource).toEqual(touchSource);
+        expect(specificSource).toEqual(touchSource2);
+
+        deviceSourceManager._removeDevice(DeviceType.Touch, 0);
+
+        const nextFirstSource = deviceSourceManager.getDeviceSource(DeviceType.Touch);
+        expect(nextFirstSource).toEqual(touchSource2);
+    });
+
+    it("can use getDeviceSources", () => {
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const touchSource = new DeviceSource(deviceInputSystem, DeviceType.Touch, 0);
+        const touchSource2 = new DeviceSource(deviceInputSystem, DeviceType.Touch, 1);
+        const emptyArray = deviceSourceManager.getDeviceSources(DeviceType.Touch);
+
+        expect(emptyArray.length).toBe(0);
+
+        deviceSourceManager._addDevice(touchSource);
+        deviceSourceManager._addDevice(touchSource2);
+
+        const touchArray = deviceSourceManager.getDeviceSources(DeviceType.Touch);
+        expect(touchArray.length).toBe(2);
+    });
+
+    it("can use onDeviceConnectedObservable", () => {
+        expect.assertions(1);
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const touchSource = new DeviceSource(deviceInputSystem, DeviceType.Touch, 0);
+        let observableSource = null;
+
+        deviceSourceManager.onDeviceConnectedObservable.add((eventData) => {
+            observableSource = eventData;
+        });
+
+        deviceSourceManager._addDevice(touchSource);
+
+        expect(observableSource).toEqual(touchSource);
+    });
+
+    it("can use onDeviceDisconnectedObservable", () => {
+        expect.assertions(2);
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const touchSource = new DeviceSource(deviceInputSystem, DeviceType.Touch, 0);
+        let observableSource = null;
+
+        deviceSourceManager.onDeviceDisconnectedObservable.add((eventData) => {

```

Comment:
`eventData` doesn't make sense here (multiple places)

```suggestion
        deviceSourceManager.onDeviceDisconnectedObservable.add((deviceSource) => {
```

Reply:
I can make that change